zerver/lib/attachments.py

```diff
@@ -146,30 +148,50 @@ def validate_attachment_request(
             usermessages_channel_ids.add(um.message.recipient.type_id)
 
     # These are subscriptions to a channel one of the messages was sent to
-    relevant_channel_ids = Subscription.objects.filter(
+    subscribed_channel_ids = Subscription.objects.filter(
         user_profile=user_profile,
         active=True,
         recipient__type=Recipient.STREAM,
         recipient__in=[m.recipient_id for m in messages],
     ).values_list("recipient__type_id", flat=True)
 
-    if usermessages_channel_ids & set(relevant_channel_ids):
+    if usermessages_channel_ids.intersection(subscribed_channel_ids):
         # If the attachment was sent in a channel with public
         # or protected history and the user is still subscribed
         # to the channel then anyone who received that message
         # can access it.
         return True, attachment
 
-    # The user didn't receive any of the messages that included this
-    # attachment. But they might still have access to it, if it was
-    # sent to a stream they are on where history is public to
-    # subscribers.
-    if len(relevant_channel_ids) == 0:
-        return False, attachment
-
-    return Stream.objects.filter(
-        id__in=relevant_channel_ids, history_public_to_subscribers=True
-    ).exists(), attachment
+    message_channel_ids = set()
+    for message in messages:
+        if message.is_stream_message():
+            message_channel_ids.add(message.recipient.type_id)
+
+    if len(message_channel_ids) > 0:
```

Comment:
this might be better written with the early return pattern -- returning if it's 0.